### ISSUE TITLE
Perl_warner isn't utf8 aware

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3513]
Migrated from [rt.perl.org#5956](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5956) (status was 'resolved')

Searchable as RT5956$

**Comment by p5pRT** [comment]
### From @nwc10



Really this is two bugs\. One's a 5\.6 or later&#8203;:

$ perl5\.6\.0 \-MDevel&#8203;::Peek \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); Dump\($a\); $a = \-$a; print $a'
SV = PV\(0x20c83fc\) at 0x20dc1b4
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\)
&nbsp;  PV = 0x20d55e0 "\\303\\260"\\0
&nbsp;  CUR = 2
&nbsp;  LEN = 5
Argument "����" isn't numeric in negation \(\-\) at \-e line 1\.
\-0
$ perl5\.6\.0 \-MDevel&#8203;::Peek \-wle '$a=v240; $a=substr\($a\,0\,1\); Dump\($a\); $a = \-$a; print $a'
SV = PV\(0x20c83fc\) at 0x20dc1ac
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\)
&nbsp;  PV = 0x20d55d8 "\\360"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
Argument "��" isn't numeric in negation \(\-\) at \-e line 1\.
\-0


surely the warning message should convert to 8 bits\, instead of showing the
raw representation? I don't think I should be able to break the utf
encapsulation even in warning messages\.

The 5\.7 bug is&#8203;:

$ perl \-MDevel&#8203;::Peek \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); Dump\($a\); $a = \-$a; print $a'
SV = PV\(0x20d975c\) at 0x20e38b8
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\,UTF8\)
&nbsp;  PV = 0x20e5470 "\\303\\260"\\0
&nbsp;  CUR = 2
&nbsp;  LEN = 5
\-��
$ perl \-MDevel&#8203;::Peek \-wle '$a=v240; $a=substr\($a\,0\,1\); Dump\($a\); $a = \-$a; print $a'
SV = PV\(0x20d975c\) at 0x20e38b0
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\)
&nbsp;  PV = 0x20e5460 "\\360"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
Argument "��" isn't numeric in negation \(\-\) at \-e line 1\.
0


where did that warning go? how did I manage to break utf8 by getting different
behaviour for the two encodings of the same character?

It's nothing to do with the IVUV "madness"\, as the lack of warning is appart
as far back as 7645 \(the earliest 5\.7\.0 I have compiled\)&#8203;:

$ \./perl \-Ilib \-MDevel&#8203;::Peek \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); Dump\($a\); $a = \-$a; print $a'
SV = PV\(0x20cde24\) at 0x20d59fc
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\,UTF8\)
&nbsp;  PV = 0x20d8c68 "\\303\\260"\\0
&nbsp;  CUR = 2
&nbsp;  LEN = 5
\-����
$ \./perl \-Ilib \-MDevel&#8203;::Peek \-wle '$a=v240; $a=substr\($a\,0\,1\); Dump\($a\); $a = \-$a; print $a'
SV = PV\(0x20cde24\) at 0x20d59f4
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\)
&nbsp;  PV = 0x20d8c60 "\\360"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
Argument "��" isn't numeric in negation \(\-\) at \-e line 1\.
\-0



Sorry that some of this is in 8 bit \- I don't think it will stop making sense
if the specific latin 1 chars get mangled\, as you can see how long the strings
are\, which is what gives the game away\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.7.0 - Fri Feb  2 23:10:57 GMT 2001
It is being executed now by  Perl v5.7.0 - Sat Feb 24 14:46:49 GMT 2001.

Site configuration information for perl v5.7.0:

Configured by nick at Sat Feb 24 14:46:49 GMT 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.0-rmk3, archname=armv4l-linux
    uname='linux bagpuss.unfortu.net 2.4.0-rmk3 #3 sat jan 20 21:38:55 gmt 2001 armv4l unknown '
    config_args='-Dmksymlinks -Dusedevel -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@talking.bollo.cx -Dperladmin=nick@talking.bollo.cx -Dinc_version_list=  -Dinc_version_list_init=0 -de'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.2 20000220 (Debian GNU/Linux)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL8670


@INC for perl v5.7.0:
    /usr/local/lib/perl5/5.7.0/armv4l-linux
    /usr/local/lib/perl5/5.7.0
    /usr/local/lib/perl5/site_perl/5.7.0/armv4l-linux
    /usr/local/lib/perl5/site_perl/5.7.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.7.0:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_GB.ISO-8859-1
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/sbin:/usr/sbin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Sat\, Mar 03\, 2001 at 10&#8203;:07&#8203;:19PM \+0000\, nick@&#8203;talking\.bollo\.cx wrote&#8203;:
> surely the warning message should convert to 8 bits\, instead of showing the
> raw representation? I don't think I should be able to break the utf
> encapsulation even in warning messages\.

Aargh\. Forgot to add&#8203;:

Is the solution to have Perl\_warner and Perl\_warner\_utf8 in the API?
\[so that we don't promote lots of things to utf8 needlessly only to downgrade
on output\, but can give utf8 warnings on utf8 aware STDERR\]

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

Nick

On Tue\, Jun 19\, 2001 at 08&#8203;:51&#8203;:44AM \-0000\, bugdb@&#8203;perl\.org wrote&#8203;:
> The status of bug\(20010303\.010\) has been updated&#8203;:
> Original status&#8203;:
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Subject&#8203;:    Perl\_warner isn't utf8 aware
> BugID  :    20010303\.010              Status&#8203;:   open
> Created&#8203;:    2001\-03\-03 14&#8203;:09&#8203;:49       Group&#8203;:    core
> Version&#8203;:    5\.7\.0\, 5\.070              Severity&#8203;: medium
> Fixed in&#8203;:                             Osname&#8203;:   linux
> Admins&#8203;:     0
> Sourceaddr&#8203;: Nicholas Clark \<nick@&#8203;talking\.bollo\.cx>
> MessageIDs&#8203;: 1   13945
> NoteIDs&#8203;:    0
> PatchIDs&#8203;:   0
> ChangeIDs&#8203;:  0
> TestIDs&#8203;:    0
> ParentID&#8203;:   0
> ChildIDs&#8203;:   0
> 
> 
> Current status&#8203;:
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Subject&#8203;:    Perl\_warner isn't utf8 aware
> BugID  :    20010303\.010              Status&#8203;:   closed
> Created&#8203;:    2001\-03\-03 14&#8203;:09&#8203;:49       Group&#8203;:    core
> Version&#8203;:    5\.7\.0                     Severity&#8203;: medium
> Fixed in&#8203;:                             Osname&#8203;:   linux
> Admins&#8203;:     0
> Sourceaddr&#8203;: Nicholas Clark \<nick@&#8203;talking\.bollo\.cx>
> MessageIDs&#8203;: 1   13945
> NoteIDs&#8203;:    0
> PatchIDs&#8203;:   0
> ChangeIDs&#8203;:  0
> TestIDs&#8203;:    0
> ParentID&#8203;:   0
> ChildIDs&#8203;:   0
> 
> To see current data on this bug\(20010303\.010\) send an email of the following form&#8203;:
> 
>     To&#8203;: bugdb@&#8203;perl\.org
>     Subject&#8203;: \-B 20010303\.010
> 
> Or to see this data on the web\, visit&#8203;:
> 
>     bugs\.perl\.org/perlbug\.cgi?req=bug\_id&bug\_id=20010303\.010
> 
> 
>

**Comment by p5pRT** [comment]
### From @jhi

The original difference between the warning being there or not has
been resolved in Perl 5\.8\.0\-\- one gets
Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.
in both cases\.
As regards to the Perl\_warner not being utf8\-aware\, I don't know\,
is there a problem?  I think if the STDOUT is UTF\-8 aware\, we can
just spit out UTF\-8?
So I'm marking the problem ticket as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Sat\, Dec 07\, 2002 at 06&#8203;:16&#8203;:52PM \-0000\, Jarkko Hietaniemi wrote&#8203;:
> The original difference between the warning being there or not has
> been resolved in Perl 5\.8\.0\-\- one gets
> Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.
> in both cases\.
> As regards to the Perl\_warner not being utf8\-aware\, I don't know\,
> is there a problem?  I think if the STDOUT is UTF\-8 aware\, we can
> just spit out UTF\-8?
> So I'm marking the problem ticket as resolved\.

The prototype is&#8203;:

void
Perl\_warn\(pTHX\_ const char \*pat\, \.\.\.\)

How does it know whether the 8 bit values I'm passing are utf8\, or latin 1
bytes that also happen to represent a valid utf8 sequence?

Nick

**Comment by p5pRT** [comment]
### From @eserte

Nicholas Clark \<nick@&#8203;unfortu\.net> writes&#8203;:

> On Sat\, Dec 07\, 2002 at 06&#8203;:16&#8203;:52PM \-0000\, Jarkko Hietaniemi wrote&#8203;:
> > The original difference between the warning being there or not has
> > been resolved in Perl 5\.8\.0\-\- one gets
> > Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.
> > in both cases\.
> > As regards to the Perl\_warner not being utf8\-aware\, I don't know\,
> > is there a problem?  I think if the STDOUT is UTF\-8 aware\, we can
> > just spit out UTF\-8?
> > So I'm marking the problem ticket as resolved\.
> 
> The prototype is&#8203;:
> 
> void
> Perl\_warn\(pTHX\_ const char \*pat\, \.\.\.\)
> 
> How does it know whether the 8 bit values I'm passing are utf8\, or latin 1
> bytes that also happen to represent a valid utf8 sequence?
> 

Is this one related?

use Test&#8203;::More 'no\_plan';
use Encode qw\(is\_utf8\);
use utf8;
$SIG\{\_\_WARN\_\_\} = sub \{ ok\(is\_utf8\($\_\[0\]\)\) \};
$x="������";
ok\(is\_utf8\($x\)\);
warn "$x\\n";
\_\_END\_\_

ok 1
not ok 2
\#     Failed test \(/tmp/u\.pl at line 19\)
1\.\.2
\# Looks like you failed 1 tests of 2\.


Regards\,
&nbsp;    Slaven

\-\- 
Slaven Rezic \- slaven\.rezic@&#8203;berlin\.de

&nbsp;    Berlin Perl Mongers \- http&#8203;://berliner\.pm\.org

**Comment by p5pRT** [comment]
### From @jhi

Ahh\, okay\, now I see the problem \(of passing in UTF\-8\)\.  Hmmm\, have
to think about this \.\. but isn't the larger problem that we can't go
around changing very call

&nbsp;    Perl\_warn\(\.\.\.\)

to

&nbsp;    SvUTF8\(somesv\) ? Perl\_warn\_utf8\(\.\.\.\) : Perl\_warn\(\.\.\.\)

In other words\, the \(char \*pat\, \.\.\.\) signature should not be
used at all\.  In other words\, how do you suggest as the solution?

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Dec 08\, 2002 at 01&#8203;:28&#8203;:31AM \-0000\, Jarkko Hietaniemi wrote&#8203;:
> Ahh\, okay\, now I see the problem \(of passing in UTF\-8\)\.  Hmmm\, have
> to think about this \.\. but isn't the larger problem that we can't go
> around changing very call
> 
>     Perl\_warn\(\.\.\.\)
> 
> to
> 
>     SvUTF8\(somesv\) ? Perl\_warn\_utf8\(\.\.\.\) : Perl\_warn\(\.\.\.\)
> 
> In other words\, the \(char \*pat\, \.\.\.\) signature should not be
> used at all\.  In other words\, how do you suggest as the solution?

Bah\. Bouncing the problem back to me :\-\)

I think we could benefit from something like

Perl\_utf8\_warn\(int is\_utf8\, const char \*pat\, \.\.\.\)

where the flag says whether the \(pattern and all the bare strings\) are treated
as utf8 \(true\) or 8 bit Latin 1 \(false\)

Clearly a 7 bit pattern doesn't care\, so if there's only 1 %s it's passing
in it sets the flag based on that\. If there are several strings being passed
in then calling code will have to check that all are compatible \(and do
conversions if needs be\)\. However\, IIRC there's a perl specific %\_ format
string that allows a pointer to an SV to be passed in\. Presumably refactoring
the core code to using this whenever possible will be the simplest way to
resolve most of the problems\, as it simplifies things \- all the utf8\-or\-not
checking is done in one place inside Perl\_warn\.

Nicholas Clark
\-\- 
INTERCAL better than perl?    http&#8203;://www\.perl\.org/advocacy/spoofathon/

**Comment by p5pRT** [comment]
### From @nwc10

\[nick \<\!\-\-c\-\-> \<i>at\</i> \<\!\-\-a\-\-> talking\.bollo\.cx \- Sat Mar  3 06&#8203;:09&#8203;:49 2001\]&#8203;:

> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> Really this is two bugs\. One's a 5\.6 or later&#8203;:
> 
> $ perl5\.6\.0 \-MDevel&#8203;::Peek \-wle '$a=v240\.257; $a=substr\($a\,0\,1\);
> Dump\($a\); $a = \-$a; print $a'
> SV = PV\(0x20c83fc\) at 0x20dc1b4
>   REFCNT = 1
>   FLAGS = \(POK\,pPOK\)
>   PV = 0x20d55e0 "\\303\\260"\\0
>   CUR = 2
>   LEN = 5
> Argument "Ã°" isn't numeric in negation \(\-\) at \-e line 1\.
> \-0
> $ perl5\.6\.0 \-MDevel&#8203;::Peek \-wle '$a=v240; $a=substr\($a\,0\,1\); Dump\($a\);
> $a = \-$a; print $a'
> SV = PV\(0x20c83fc\) at 0x20dc1ac
>   REFCNT = 1
>   FLAGS = \(POK\,pPOK\)
>   PV = 0x20d55d8 "\\360"\\0
>   CUR = 1
>   LEN = 2
> Argument "ð" isn't numeric in negation \(\-\) at \-e line 1\.
> \-0
> 
> 
> surely the warning message should convert to 8 bits\, instead of
> showing the
> raw representation? I don't think I should be able to break the utf
> encapsulation even in warning messages\.
> 
> The 5\.7 bug is&#8203;:
> 
> $ perl \-MDevel&#8203;::Peek \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); Dump\($a\);
> $a = \-$a; print $a'
> SV = PV\(0x20d975c\) at 0x20e38b8
>   REFCNT = 1
>   FLAGS = \(POK\,pPOK\,UTF8\)
>   PV = 0x20e5470 "\\303\\260"\\0
>   CUR = 2
>   LEN = 5
> \-ð
> $ perl \-MDevel&#8203;::Peek \-wle '$a=v240; $a=substr\($a\,0\,1\); Dump\($a\); $a =
> \-$a; print $a'
> SV = PV\(0x20d975c\) at 0x20e38b0
>   REFCNT = 1
>   FLAGS = \(POK\,pPOK\)
>   PV = 0x20e5460 "\\360"\\0
>   CUR = 1
>   LEN = 2
> Argument "ð" isn't numeric in negation \(\-\) at \-e line 1\.
> 0
> 
> 
> where did that warning go? how did I manage to break utf8 by getting
> different
> behaviour for the two encodings of the same character?
> 
> It's nothing to do with the IVUV "madness"\, as the lack of warning is
> appart
> as far back as 7645 \(the earliest 5\.7\.0 I have compiled\)&#8203;:
> 
> $ \./perl \-Ilib \-MDevel&#8203;::Peek \-wle '$a=v240\.257; $a=substr\($a\,0\,1\);
> Dump\($a\); $a = \-$a; print $a'
> SV = PV\(0x20cde24\) at 0x20d59fc
>   REFCNT = 1
>   FLAGS = \(POK\,pPOK\,UTF8\)
>   PV = 0x20d8c68 "\\303\\260"\\0
>   CUR = 2
>   LEN = 5
> \-Ã°
> $ \./perl \-Ilib \-MDevel&#8203;::Peek \-wle '$a=v240; $a=substr\($a\,0\,1\);
> Dump\($a\); $a = \-$a; print $a'
> SV = PV\(0x20cde24\) at 0x20d59f4
>   REFCNT = 1
>   FLAGS = \(POK\,pPOK\)
>   PV = 0x20d8c60 "\\360"\\0
>   CUR = 1
>   LEN = 2
> Argument "ð" isn't numeric in negation \(\-\) at \-e line 1\.
> \-0

The latter bug is still present on 5\.6\.1&#8203;:

$ perl5\.6\.1 \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); $a = \-$a;'
$ perl5\.6\.1 \-wle '$a=v240; $a=substr\($a\,0\,1\); $a = \-$a;'
Argument "M\-p" isn't numeric in negation \(\-\) at \-e line 1\.

The former bug is fixed by 5\.8\.0\, at least for the case of the numeric conversion warning\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @gannett-ggreer

On Sun Aug 24 11&#8203;:25&#8203;:14 2003\, nicholas wrote&#8203;:
> The latter bug is still present on 5\.6\.1&#8203;:
> 
> $ perl5\.6\.1 \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); $a = \-$a;'
> $ perl5\.6\.1 \-wle '$a=v240; $a=substr\($a\,0\,1\); $a = \-$a;'
> Argument "M\-p" isn't numeric in negation \(\-\) at \-e line 1\.
> 
> The former bug is fixed by 5\.8\.0\, at least for the case of the numeric
> conversion warning\.

The perl 5\.10\.1 that comes with my Ubuntu Linux as well as standard
5\.13\.3 and 5\.12\.1 are printing for those two lines&#8203;:

Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.

http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=5956

\-\- 
George Greer

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jul 28 20&#8203;:23&#8203;:36 2010\, greerga wrote&#8203;:
> On Sun Aug 24 11&#8203;:25&#8203;:14 2003\, nicholas wrote&#8203;:
> > The latter bug is still present on 5\.6\.1&#8203;:
> > 
> > $ perl5\.6\.1 \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); $a = \-$a;'
> > $ perl5\.6\.1 \-wle '$a=v240; $a=substr\($a\,0\,1\); $a = \-$a;'
> > Argument "M\-p" isn't numeric in negation \(\-\) at \-e line 1\.
> > 
> > The former bug is fixed by 5\.8\.0\, at least for the case of the numeric
> > conversion warning\.
> 
> The perl 5\.10\.1 that comes with my Ubuntu Linux as well as standard
> 5\.13\.3 and 5\.12\.1 are printing for those two lines&#8203;:
> 
> Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.
> 
> http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=5956

Does that mean this can be marked as resolved?

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Dec 07 14&#8203;:31&#8203;:20 2002\, nick@&#8203;unfortu\.net wrote&#8203;:
> On Sat\, Dec 07\, 2002 at 06&#8203;:16&#8203;:52PM \-0000\, Jarkko Hietaniemi wrote&#8203;:
> > The original difference between the warning being there or not has
> > been resolved in Perl 5\.8\.0\-\- one gets
> > Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.
> > in both cases\.
> > As regards to the Perl\_warner not being utf8\-aware\, I don't know\,
> > is there a problem?  I think if the STDOUT is UTF\-8 aware\, we can
> > just spit out UTF\-8?
> > So I'm marking the problem ticket as resolved\.
> 
> The prototype is&#8203;:
> 
> void
> Perl\_warn\(pTHX\_ const char \*pat\, \.\.\.\)
> 
> How does it know whether the 8 bit values I'm passing are utf8\, or latin 1
> bytes that also happen to represent a valid utf8 sequence?

I don’t know when it was added\, but we have SVf for precisely this\.  In
5\.15\.4\, almost all error messages with symbol names were changed to use
SVf or the new HEKf instead of %s\, so I think this ticket can be resolved\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Sun Sep 12 18&#8203;:09&#8203;:11 2010\, sprout wrote&#8203;:
> On Wed Jul 28 20&#8203;:23&#8203;:36 2010\, greerga wrote&#8203;:
> > On Sun Aug 24 11&#8203;:25&#8203;:14 2003\, nicholas wrote&#8203;:
> > > The latter bug is still present on 5\.6\.1&#8203;:
> > > 
> > > $ perl5\.6\.1 \-wle '$a=v240\.257; $a=substr\($a\,0\,1\); $a = \-$a;'
> > > $ perl5\.6\.1 \-wle '$a=v240; $a=substr\($a\,0\,1\); $a = \-$a;'
> > > Argument "M\-p" isn't numeric in negation \(\-\) at \-e line 1\.
> > > 
> > > The former bug is fixed by 5\.8\.0\, at least for the case of the numeric
> > > conversion warning\.
> > 
> > The perl 5\.10\.1 that comes with my Ubuntu Linux as well as standard
> > 5\.13\.3 and 5\.12\.1 are printing for those two lines&#8203;:
> > 
> > Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.
> > 
> > http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=5956
> 
> Does that mean this can be marked as resolved?

Well\, re\-resolved\, as Jarkko originally marked it as resolved\,
when he made this change to fix it&#8203;:

commit 8eb28a70b2ec19f2782a68fd1ccf1a9a24131140
Author&#8203;: Jarkko Hietaniemi \<jhi@&#8203;iki\.fi>
Date&#8203;:   Tue Oct 23 22&#8203;:19&#8203;:34 2001 \+0000

&nbsp;    Negation and Unicode&#8203;: sort of solves 20010303\.010\,
&nbsp;    except not quite like reported in the Subject
&nbsp;    \(Perl\_warner is still utf8\-ignorant\)\.
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;12614

 pp\.c              |   21 \+\+\+\+\+\+\+\+\+\+\+\+\+\+\-\-\-\-\-\-\-
 t/lib/warnings/sv |    8 \+\+\+\+\+\+\+\+
 2 files changed\, 22 insertions\(\+\)\, 7 deletions\(\-\)


What's strange is that prior to that commit\, post 5\.6\.0\, the warning
had been eliminated\. Turns out to have been by this seemginly
unrelated commit&#8203;:

commit 3bd709b1a63d554f3d98d5394be78ed628eb46da
Author&#8203;: Peter Prymmer \<PPrymmer@&#8203;factset\.com>
Date&#8203;:   Thu Mar 8 08&#8203;:23&#8203;:25 2001 \-0800

&nbsp;    Re&#8203;: Unicode/EBCDIC
&nbsp;    Message\-ID&#8203;:
\<Pine\.OSF\.4\.10\.10103081617390\.377472\-100000@&#8203;aspara\.forte\.com>
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;9082

 perl\.c |    6 \+\-
 perl\.h |  216
\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\-
 sv\.c   |    4 \+\+
 toke\.c |   10 \+\+\-
 utf8\.c |    4 \+\-
 utf8\.h |   19 \+\+\+\+\+\+
 6 files changed\, 251 insertions\(\+\)\, 8 deletions\(\-\)

which doesn't \*seem\* to touch anything in the relevant code paths\.
\(But I have checked manually \- it did do it\. Yet more strange
action at distance\. Crazy codebase\.\)

On Wed Jan 11 09&#8203;:58&#8203;:17 2012\, sprout wrote&#8203;:
> On Sat Dec 07 14&#8203;:31&#8203;:20 2002\, nick@&#8203;unfortu\.net wrote&#8203;:
> > On Sat\, Dec 07\, 2002 at 06&#8203;:16&#8203;:52PM \-0000\, Jarkko Hietaniemi wrote&#8203;:
> > > The original difference between the warning being there or not has
> > > been resolved in Perl 5\.8\.0\-\- one gets
> > > Argument "\\x\{f0\}" isn't numeric in negation \(\-\) at \-e line 1\.
> > > in both cases\.
> > > As regards to the Perl\_warner not being utf8\-aware\, I don't know\,
> > > is there a problem?  I think if the STDOUT is UTF\-8 aware\, we can
> > > just spit out UTF\-8?
> > > So I'm marking the problem ticket as resolved\.
> > 
> > The prototype is&#8203;:
> > 
> > void
> > Perl\_warn\(pTHX\_ const char \*pat\, \.\.\.\)
> > 
> > How does it know whether the 8 bit values I'm passing are utf8\, or
latin 1
> > bytes that also happen to represent a valid utf8 sequence?
> 
> I don’t know when it was added\, but we have SVf for precisely this\.  In
> 5\.15\.4\, almost all error messages with symbol names were changed to use
> SVf or the new HEKf instead of %s\, so I think this ticket can be resolved\.

The macro SVf was added in 2000&#8203;:

commit 894356b32151f778d4d2915c6db38e5d049b115a
Author&#8203;: Gurusamy Sarathy \<gsar@&#8203;cpan\.org>
Date&#8203;:   Sat Jan 22 10&#8203;:06&#8203;:53 2000 \+0000

&nbsp;    add patch for printf\-style format typechecks \(from Robin Barker
&nbsp;    \<rmb1@&#8203;cise\.npl\.co\.uk>\); fixes for problems so identified
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;4836

but I know that the definition of a format for SVs goes back much
further\.\.\.

5\.004 it seems \(commit fc36a67e8855d031\, in April 1997\)

However\, I think that the historic problem was that a lot of code
called /warn/ and /croak/ functions with char \* pointers\, and at
the time Jarkko and I were assuming that the logical fix was to have
some way of saying "char \* but UTF\-8"\. Probably partly because we'd
not thought of SVf\, but I'm guessing also that because in at least
some places \(at the point of calling /warn/ or /croak/\) the value
about to be complained about was in a char \*\.

I'm guessing that these days\, after a lot of UTF\-8 cleanup\, the value
is often already in an SV that has been passed down far enough\, as
half the problem \(alluded to in parens above\) is that the
"UTF\-8 or not" had already been lost earlier in the C call stack\,
when something naively used SvPV\(\) without caring about SvUTF8\(\)\.

So yes\, both problems referenced in this bug are resolved\.
I'm sure specific issue with UTF\-8 wrongness remain\, but thanks to
you\, Brian\, Karl and others there are far far fewer\. And any little
critters still existing deserve their own tickets\.

Nicholas Clark
